src/stirling/source_connectors/socket_tracer/protocols/mongodb/stitcher.cc

```diff
@@ -0,0 +1,210 @@
+    mongodb::Frame& next_resp = *next_resp_it;
+    resp_frame->sections.insert(std::end(resp_frame->sections), std::begin(next_resp.sections),
+                                std::end(next_resp.sections));
+    next_resp.consumed = true;
+    *latest_resp_ts = next_resp.timestamp_ns;
+    curr_resp = &next_resp;
+  }
+
+  // TODO(kpattaswamy): In the case of "missing" more to come middle/tail frames, determine whether
+  // they are truly missing or have not been parsed yet.
+}
+
+void FlattenSections(mongodb::Frame* frame) {
+  // Flatten the vector of sections containing vector of documents into a single string.
+  for (const auto& section : frame->sections) {
+    for (const auto& doc : section.documents) {
+      frame->frame_body.append(doc).append(" ");
+    }
+  }
+  frame->sections.clear();
+}
+
+RecordsWithErrorCount<mongodb::Record> StitchFrames(
+    absl::flat_hash_map<mongodb::stream_id_t, std::deque<mongodb::Frame>>* reqs,
+    absl::flat_hash_map<mongodb::stream_id_t, std::deque<mongodb::Frame>>* resps, State* state) {
+  std::vector<mongodb::Record> records;
+  int error_count = 0;
+
+  for (auto stream_id_it = state->stream_order.begin(); stream_id_it != state->stream_order.end();
+       stream_id_it++) {
+    auto& stream_id_pair = *stream_id_it;
+    auto stream_id = stream_id_pair.first;
+
+    // Find the stream ID's response deque.
+    auto resp_it = resps->find(stream_id);
+    if (resp_it == resps->end()) {
+      VLOG(1) << absl::Substitute("Did not find a response deque with the stream ID: $0",
+                                  stream_id);
+      continue;
+    }
+
+    // Response deque for the stream ID.
+    auto& resp_deque = resp_it->second;
+
+    // Find the stream ID's request deque.
+    auto req_it = reqs->find(stream_id);
+    // The request deque should exist in the reqs map since the state contained the stream ID.
+    CTX_DCHECK(req_it != reqs->end());
+
+    // Request deque for the stream ID.
+    auto& req_deque = req_it->second;
+
+    // Track the latest response timestamp to compare against request frame's timestamp later.
+    uint64_t latest_resp_ts = 0;
+
+    // Stitch the first frame in the response deque with the corresponding request frame.
+    for (const auto& [idx, resp_frame] : Enumerate(resp_deque)) {
+      if (resp_frame.consumed) {
+        continue;
+      }
+
+      latest_resp_ts = resp_frame.timestamp_ns;
+
+      // Find the corresponding request frame for the head response frame.
+      auto req_frame_it = std::upper_bound(req_deque.begin(), req_deque.end(), latest_resp_ts,
+                                           [](const uint64_t ts, const mongodb::Frame& frame) {
+                                             return ts < frame.timestamp_ns;
+                                           }) -
+                          1;
+      if (req_frame_it == req_deque.begin() &&
+          req_frame_it->timestamp_ns > resp_frame.timestamp_ns) {
```

Comment:
@kpattaswamy have you validated that this condition works? If there are no earlier requests, I think we will get `req_deque.begin() - 1` because we are subtracting 1 from the `std::upper_bound` output.